### PR TITLE
Return success status of grpc_byte_buffer_reader_init

### LINKED ISSUE #7149: Decompression error is not captured and propagated to upper layers
When the decompression algorithm fails, the current decompression code directly pass compressed data to user. May need to add error report and handing in this part of code.

Code ref: https://github.com/grpc/grpc/blob/1f6faf3725453d45953b2a55feafd7a368f96fff/src/core/lib/surface/byte_buffer_reader.c#L68-L72

### PR DESCRIPTION
**Pull request by dgquintas**
`grpc_byte_buffer_reader_init` now returns an `int` (effectively a bool) indicating errors.

I've changed the wrapping code in objective-c, node, ruby and php to the best of my knowledge. Changes are still needed for Python and C# to consider `grpc_byte_buffer_reader_init` and react to failures appropriately. I've added a TODO for those two languages.

@apolcyn for C#
@murgatroid99 for Ruby and Node.
@makdharma / @jcanizales for obj-c
@nathanielmanistaatgoogle / @soltanmm for Python.
@yang-g for C++
@stanley-cheung for PHP
@ctiller for core.

Fixes #7149

### CODE REVIEW

File: src/ruby/ext/grpc/rb_byte_buffer.c [hunk 1]
```diff
@@ -56,7 +56,10 @@ VALUE grpc_rb_byte_buffer_to_s(grpc_byte_buffer *buffer) {
     return Qnil;
   }
   rb_string = rb_str_buf_new(grpc_byte_buffer_length(buffer));
-  grpc_byte_buffer_reader_init(&reader, buffer);
+  if (!grpc_byte_buffer_reader_init(&reader, buffer)) {
+    rb_raise(rb_eRuntimeError, error_message);
```
Comment (murgatroid99) [inline src/ruby/ext/grpc/rb_byte_buffer.c]: "error_message" was just a placeholder. You need to write out an error message like in the other cases.

File: src/objective-c/GRPCClient/private/NSData+GRPC.m [hunk 2]
```diff
@@ -42,7 +42,11 @@
 static void MallocAndCopyByteBufferToCharArray(grpc_byte_buffer *buffer,
                                                size_t *length, char **array) {
   grpc_byte_buffer_reader reader;
-  grpc_byte_buffer_reader_init(&reader, buffer);
+  if (!grpc_byte_buffer_reader_init(&reader, buffer)) {
```
Comment (murgatroid99) [inline src/objective-c/GRPCClient/private/NSData+GRPC.m]: Please add a TODO about checking the return code.

File: src/php/ext/grpc/byte_buffer.c [hunk 3]
```diff
@@ -58,16 +58,15 @@ grpc_byte_buffer *string_to_byte_buffer(char *string, size_t length) {
 
 void byte_buffer_to_string(grpc_byte_buffer *buffer, char **out_string,
                            size_t *out_length) {
-  if (buffer == NULL) {
+  grpc_byte_buffer_reader reader;
+  if (buffer == NULL || !grpc_byte_buffer_reader_init(&reader, buffer)) {
```
Comment (murgatroid99) [inline src/php/ext/grpc/byte_buffer.c]: Please add a TODO about checking the return code.

File: src/ruby/ext/grpc/rb_byte_buffer.c (hunk 1 above)
Comment (dgquintas) [inline src/ruby/ext/grpc/rb_byte_buffer.c]: Woops. Fixed.

File: src/php/ext/grpc/byte_buffer.c (hunk 3 above)
Comment (stanley-cheung) [inline src/php/ext/grpc/byte_buffer.c]: Agreed with TODO. otherwise LGTM

Comment (dgquintas) [inline src/php/ext/grpc/byte_buffer.c]: Done.

File: src/objective-c/GRPCClient/private/NSData+GRPC.m (hunk 2 above)
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/NSData+GRPC.m]: Please add a comment explaining that `grpc_byte_buffer_reader_init` can fail if the data sent by the server could not be decompressed for any reason; that the server is to blame; and thus we want the RPC to fail with error code `INTERNAL`.

Comment (dgquintas) [inline src/objective-c/GRPCClient/private/NSData+GRPC.m]: That seems to expose internal details that may not be true in the future. All this code should care about is that the initialization of the reader failed. Arguably, we may want to add some function (eg, `grpc_byte_buffer_reader_status`) expanding on the reason for the failure. How does that sound?

Comment (dgquintas) [inline src/objective-c/GRPCClient/private/NSData+GRPC.m]: Actually, nevermind... let's not over-engineer. Adding the comment...

Comment (dgquintas) [inline src/objective-c/GRPCClient/private/NSData+GRPC.m]: Done.

**Comment by yang-g** [comment]
LGTM

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi [hunk 4]
```diff
@@ -254,6 +254,7 @@ cdef class ByteBuffer:
     cdef void *data_slice_pointer
     if self.c_byte_buffer != NULL:
       with nogil:
+        # TODO(dgq): check the return value of grpc_byte_buffer_reader_init.
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: `TODO(<link to filed issue>)` rather than `TODO(username)`, please.

**Comment by murgatroid99** [comment]
LGTM

**Comment by makdharma** [comment]
LGTM

File: src/core/lib/surface/byte_buffer_reader.c [hunk 5]
```diff
@@ -69,7 +69,8 @@ void grpc_byte_buffer_reader_init(grpc_byte_buffer_reader *reader,
                   "Unexpected error decompressing data for algorithm with enum "
                   "value '%d'. Reading data as if it were uncompressed.",
                   reader->buffer_in->data.raw.compression);
-          reader->buffer_out = reader->buffer_in;
+          return 0;
```
Comment (ctiller) [inline src/core/lib/surface/byte_buffer_reader.c]: Probably the memset should come before the return?

**Comment by jcanizales** [comment]
ObjC change LGTM

**Comment by apolcyn** [comment]
LGTM

File: src/core/lib/surface/byte_buffer_reader.c (hunk 5 above)
Comment (dgquintas) [inline src/core/lib/surface/byte_buffer_reader.c]: Done

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi (hunk 4 above)
Comment (dgquintas) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: Done

**Comment by dgquintas** [comment]
@ctiller & @nathanielmanistaatgoogle PTAL.

**Comment by nathanielmanistaatgoogle** [comment]
1. If you wanted to just fold @soltanmm's code from [his follow-up pull request](https://github.com/grpc/grpc/pull/7213/files) into this pull request, that would be fine.
2. What hope is there of test coverage?

**Comment by dgquintas** [comment]
1. Ack. We are trying to be fancy and have @soltanmm commit against my branch, but github is hard. Failing that I'll just replicate his changes manually.
2. All the hope. Will do.

**Comment by nathanielmanistaatgoogle** [comment]
Oh. I was expecting you to say "none", and then I was going to say "well that's okay, please describe the impossibility of test coverage very thoroughly in your commit log message". But actual test coverage would be even better.

**Comment by dgquintas** [comment]
Well, it's a matter of creating a corrupted byte buffer and feeding it to the function. Were you referring to something other than a C unit test for the newly add functionality in `byte_buffer_reader.c`?

**Comment by nathanielmanistaatgoogle** [comment]
Yeah, I was specifically referring to test coverage of the changed Python code.

**Comment by dgquintas** [comment]
I think the same principle would apply to Python. I'll write the C test and then @soltanmm could perhaps update https://github.com/grpc/grpc/pull/7213 with the corresponding test, which if I'm not mistaken would be very similar to the C one.

**Comment by dgquintas** [comment]
See https://github.com/grpc/grpc/pull/7190/commits/ecbe2d52853ee70716479f0eaaef8564667255ec#diff-fd7344b886003cac58aafbbc4bdacb54R118 for the test. I presume replicating this in Python is feasible? @soltanmm :grimacing: ?

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* One or more commits are not associated with a GitHub user. The email used in the Git commit must match a verified email address in a GitHub user profile. For more information, please see https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/
* User @soltanmm-google isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Pull request merged**